doc/development_environment.md
# Development Environment

## Install Compilers and Build Tools

- Before anything else, do note that **an antikythera instance and its gears run in the same ErlangVM cluster**.
  Therefore, obviously, all development AND deployment environments must have the same Erlang and Elixir versions!
- We strongly recommend [asdf](https://github.com/asdf-vm/asdf) to manage this synchronization.
    - See [antikythera's `.tool-versions` file](https://github.com/access-company/antikythera/blob/master/.tool-versions)
      for the versions of Erlang and Elixir supported in current revision.
        - Antikythera itself has strict version checking mechanism and raises on compile if different versions are used in your environment.
        - (This may change later so that Erlang and Elixir versions can be checked by individual antikythera instances, not by antikythera)
- Other requirements:
    - Some dependencies contain C source code and thus a reasonably new C compiler is required.
    - In macOS you may be warned by outdated GNU make version during compiling antikythera's dependencies.
      In that case install GNU make using e.g. [homebrew](http://brew.sh/).
      The installed executable is `gmake`; create a symlink named `make` in somewhere in your `$PATH`
      so that the desired version of make command is selected.
    - Install development header of [Expat](https://libexpat.github.io/), which is required by [fast_xml](https://github.com/processone/fast_xml) library.
        - In macOS this should be already installed. If not, install `expat` with homebrew.
    - On Linux you have to install `inotify-tools` to enable file system watching tools
      such as [exsync](https://github.com/falood/exsync) and [mix_test_watch](https://github.com/lpil/mix-test.watch).
- [Install asdf as explained in the official README.md](https://github.com/asdf-vm/asdf#setup)
- Install asdf plugins
    - `$ asdf plugin-add erlang`
    - `$ asdf plugin-add elixir`
    - `$ asdf plugin-add nodejs`
        - Node.js installation is optional. You do not need it if you are not using asset-related features.
- Install language versions
    - In order to invoke initial `mix deps.get`, you need to globally install reasonably new Erlang and Elixir versions.
        - `$ asdf install erlang x.y.z && asdf global erlang x.y.z`
        - `$ asdf install elixir x.y.z && asdf global elixir x.y.z`
    - Working on your antikythera instance:
        - If `.tool-versions` already exists, `$ asdf install`
        - If not,
            - `$ asdf local erlang x.y.z && asdf local elixir x.y.z`
            - then commit the generated `.tool-versions` file.
    - Working on a gear:
        - Make sure you have a valid symlink named `.tool-versions` which points to `deps/your_antikythera_instance/.tool-versions`
            - This way you can make your gear's language versions in sync with your antikythera instance.
            - If your gear is generated using [`mix antikythera.gear.new`](https://github.com/access-company/antikythera/blob/master/lib/mix/gear.new.ex),
              symlink to `.tool-versions` is automatically created.
        - `$ asdf install`


### Shell history in `iex`

- Although `iex` is able to remember commands executed in previous sessions, the feature is not enabled by default.
- It is highly recommended to set the following environment variable during development:
    - `export ERL_AFLAGS="-kernel shell_history enabled"` (for bash family)

## Prepare for Domain-based Routing

- Antikythera routes web requests by subdomains and paths. See [Routing](./gear_developers/routing.md).
- In local development, you must be able to resolve subdomain of localhost (e.g. `your-gear.localhost`) into loopback address (`127.0.0.1`).
    - The easiest way is to add a line like `127.0.0.1 your-gear.localhost` to your `/etc/hosts`.
    - Alternatively, you can setup a local DNS server (such as [dnsmasq](http://www.thekelleys.org.uk/dnsmasq/doc.html)) to resolve all subdomains of localhost.

## Environment variables to tweak behavior of antikythera

- In your local environment you can customize some of antikythera's default settings using the following environment variables:
    - Runtime environment variables:
        - `BASE_DOMAIN`:
            - Suffix part of domain names to be used when clients interact with running gears. Defaults to `localhost`.
            - For example, when you are running `:some_gear` as `BASE_DOMAIN=somedomain iex -S mix`,
              you can browse the top page of the gear at `http://some-gear.somedomain:8080/` (assuming that the default port number is used; see below).
            - Note that when using `BASE_DOMAIN` other than `localhost`, you may need to tell the DNS resolver on the client side about the gear's domain name.
        - `PORT`:
            - Port number to receive incoming web requests from. Defaults to `8081` during `mix test`, and `8080` otherwise
              (thus one can run both `iex -S mix` and `mix test` at the same time).
            - Explicitly set this when you run multiple antikythera servers within your machine.
        - `TEST_PORT`:
            - Port number to send HTTP requests to during `blackbox_local` tests. Defaults to `8080`.
        - `NO_LISTEN`:
            - A boolean flag for disabling web server functionality of antikythera. Defaults to `false`.
            - `true` is implicitly set when using `Antikythera.Mix.Task.prepare_antikythera/0` in your mix task.
        - `LOG_LEVEL`:
            - Log level of all gears. Defaults to `info`. See also [logging.md](./gear_developers/logging.md).
        - `SOME_GEAR_CONFIG_JSON`:
            - Gear config of `:some_gear`. See also [gear_config.md](./gear_developers/gear_config.md).
    - Compile-time environment variables: (to change the followings you need to recompile antikythera)
        - `GEAR_ACTION_TIMEOUT`:
            - Milliseconds to wait until gear action finishes. Defaults to `10000`. See also [controller.md](./gear_developers/controller.md).
        - `GEAR_PROCESS_MAX_HEAP_SIZE`:
            - Maximum size of per-process heap memory in words. Defaults to `50000000` (400MB in 64-bit architecture).


doc/gear_developers/gear_config.md
# Gear Config

**Note:** This page is being updated for OSS release. Please be patient.

- In general web services need some kind of "secret token"s, such as API key for external services.
  Also, it's sometimes useful to change configurations on the fly (i.e. without changing source code and without deployment).
- For these purposes antikythera provides "gear config".
    - You can think of gear config as something like environment variables for gears.
    - Each gear's gear config is an arbitrary JSON object.

## Getting gear config

- From gear implementation, gear config is accessble by:
    - `YourGear.get_all_env/0`: Returns a map of JSON-parsed gear config.
    - `YourGear.get_env/2`: A convenience to just return a single value for the given key in gear config.
      Use this when you just want a single value; if you need multiple values use `get_all_env/0` for better performance.

## Setting gear config

- In cloud (dev/prod environment):
    - **(TBD)**
    - After update, new values of gear config become visible from gear code, running within multiple ErlangVM nodes, in about a few minutes.
- In local machine
    - Use environment variable to specify gear config for your locally-running antikythera.
      For instance, set `SOME_GEAR_CONFIG_JSON` for `some_gear` and so on.
- During tests:
    - You can use `Antikythera.Test.GearConfigHelper.set_config/1` to manipulate gear config from test code.


CHANGELOG.md
# Changelog of Interfaces Provided by Antikythera

This file contains the history of **backward-incompatible changes** in interfaces of antikythera
which are exposed to antikythera instance administrators and/or gear developers.

---


lib/util/env.ex
# Copyright(c) 2015-2018 ACCESS CO., LTD. All rights reserved.

use Croma

defmodule Antikythera.Env do
  @moduledoc """
  Module to provide helpers to access environment variables defined by antikythera.

  ## Environments and deployments

  Antikythera instances and gears may run in the following two modes:

  - As a mix project, when invoked by running `iex` or `mix` command-line tool.
  - As an OTP release, which is generated by e.g. `antikythera_core.generate_release` task.

  In general there are multiple deployments per antikythera instance.
  To distingish the target deployment from code, antikythera defines the following environment variables:

  - At compile time (for metaprogramming):
    `ANTIKYTHERA_COMPILE_ENV` must be appropriately set; the value can be retrieved by `compile_env/0`.
  - At runtime:
    `ANTIKYTHERA_RUNTIME_ENV` must be appropriately set; the value can be retrieved by `runtime_env/0`.

  It is the responsibility of antikythera instance administrators to correctly set these environment variables
  when compiling/running antikythera and gears.
  Possible values returned by `compile_env/0` and `runtime_env/0` are:

  - names of deployments given by `:deployments` application config
  - `:local` (running an OTP release at local machine for testing purpose)
  - `:undefined`

  As an example, if you set `:dev` and `:prod` in `:deployments`, then you get:

      |                           | Mix.env() at compile-time              | Mix.env() at runtime    | compile_env(), runtime_env() |
      |---------------------------+----------------------------------------+-------------------------+------------------------------|
      | $ iex -S mix              | :dev  (toplevel), :prod (dependencies) | :dev                    | :undefined                   |
      | $ mix test                | :test (toplevel), :prod (dependencies) | :test                   | :undefined                   |
      | $ mix antikythera_local.* | :prod                                  | (:mix is not available) | :local                       |
      | :dev deployment           | :prod                                  | (:mix is not available) | :dev                         |
      | :prod deployment          | :prod                                  | (:mix is not available) | :prod                        |

  You can use these values to distinguish the current context from your code.
  """

  deployment_envs = Application.fetch_env!(:antikythera, :deployments) |> Keyword.keys()
  use Croma.SubtypeOfAtom, values: (deployment_envs ++ [:local, :undefined])
  alias Antikythera.{GearName, GearNameStr, Url}
  alias AntikytheraCore.Handler.CowboyRouting, as: Routing

  defmodule Mapping do
    Enum.each(deployment_envs, fn env ->
      env_str = Atom.to_string(env)
      def env_var_to_atom(unquote(env_str)), do: unquote(env)
    end)
    def env_var_to_atom("local"), do: :local
    def env_var_to_atom(_      ), do: :undefined

    Enum.each(deployment_envs, fn env ->
      def cloud?(unquote(env)), do: true
    end)
    def cloud?(_), do: false
  end

  defun no_listen?() :: boolean do
    case System.get_env() do
      %{"NO_LISTEN" => "true"}           -> true
      %{"TEST_MODE" => "blackbox_" <> _} -> true
      _otherwise                         -> false
    end
  end

  defun runtime_env()           :: t      , do: System.get_env("ANTIKYTHERA_RUNTIME_ENV") |> Mapping.env_var_to_atom()
  defun running_with_release?() :: boolean, do: runtime_env() != :undefined
  defun running_in_cloud?()     :: boolean, do: Mapping.cloud?(runtime_env())

  @compile_env            System.get_env("ANTIKYTHERA_COMPILE_ENV") |> Mapping.env_var_to_atom()
  @compiling_for_release? @compile_env != :undefined
  @compiling_for_cloud?   Mapping.cloud?(@compile_env)

  defun compile_env()            :: t      , do: @compile_env
  defun compiling_for_release?() :: boolean, do: @compiling_for_release?
  defun compiling_for_cloud?()   :: boolean, do: @compiling_for_cloud?

  @antikythera_instance_name Application.fetch_env!(:antikythera, :antikythera_instance_name)
  defun antikythera_instance_name() :: atom, do: @antikythera_instance_name

  @gear_action_timeout_default 10_000
  @gear_action_timeout         String.to_integer(System.get_env("GEAR_ACTION_TIMEOUT") || "#{@gear_action_timeout_default}")
  @doc """
  Timeout (in milli-seconds) for gear actions.

  This can be configurable by specifying `"GEAR_ACTION_TIMEOUT"` environment variable when compiling antikythera.
  Defaults to `#{@gear_action_timeout_default}`.
  """
  defun gear_action_timeout() :: pos_integer, do: @gear_action_timeout

  @default_port             8080
  @default_port_during_test 8081

  # To see which port number to use, we have to get the current mix environment at runtime,
  # since `Mix.env()` always returns `:prod` during compilation of antikythera within a gear project.
  # However, when running with an OTP release, `Mix.env/0` is not available at runtime
  # (as `:mix` is not included in antikythera's runtime dependencies).
  # Therefore we need both compile-time- and runtime-branching.
  if @compiling_for_release? do
    defp default_port_at_runtime(), do: @default_port
  else
    defp default_port_at_runtime(), do: (if Mix.env() == :test, do: @default_port_during_test, else: @default_port)
  end

  @doc """
  TCP port to listen to for incoming web requests.

  The port can be specified by `"PORT"` runtime environment variable.
  Defaults to `#{@default_port_during_test}` during `mix test`, and `#{@default_port}` otherwise
  (thus one can run both `iex -S mix` and `mix test` at the same time).
  """
  defun port_to_listen() :: non_neg_integer do
    case System.get_env("PORT") do
      nil      -> default_port_at_runtime()
      port_str -> String.to_integer(port_str)
    end
  end

  defun default_base_url(gear_name :: v[GearName.t | GearNameStr.t], env :: v[t] \\ @compile_env) :: Url.t do
    if Mapping.cloud?(env) do
      "https://" <> Routing.default_domain(gear_name, env)
    else
      "http://" <> Routing.default_domain(gear_name, env) <> ":#{port_to_listen()}"
    end
  end
end
